### PR TITLE
rgw: add `rgw-orphan-list` tool & `radosgw-admin bucket radoslist ...`

### PR DESCRIPTION
**Pull request by ivancich**
Adds a radosgw-admin subcommand and possibly a bucket name (if not, all buckets are assumed) and walks the bucket(s) listing(s) and the manifest (if it exists) for each object in the listing to generate the rados objects that represent the rgw objects in the bucket.
    
Also adds a tool named rgw-orphan-list that will produce a list in a local file of what appear to be rgw orphans.

Tracker: https://tracker.ceph.com/issues/44733

### CODE REVIEW

**Comment by ivancich** [comment]
@tchaikov and @smithfarm : please review to make sure build is correct (and more if you choose!)

**Comment by smithfarm** [comment]
@ivancich Are these commits cherry-picks from the nautilus PR? There is an expectation that, if a backward- and forward-ported commit is very similar (or identical) to an existing commit, the commit message will somehow mention that.

Also, would it make sense to re-target this PR to master, get it merged, and then cherry-pick backwards from there? (Just talking about the octopus PR here, given that octopus and master are very similar, I'm trying to understand how forward-port is justified in this case.)

**Comment by ivancich** [comment]
@smithfarm We could pretend it started at master and was backported. In truth, the nautilus version came across first as it was closest to downstream. And I cherry-picked it upwards to octopus/master from there.

And yes, I'll retarget to master.

It's up to you. Should we start with master and pretend we cherry-picked and backported from there?

**Comment by ivancich** [comment]
Even though the underlying branch name contains "octopus", this is targeted at master. I'd rather not close this pr and create one with a better branch name in order to preserve the conversation taking place.

**Comment by smithfarm** [comment]
> Should we start with master and pretend we cherry-picked and backported from there?

I converted the nautilus tracker issue https://tracker.ceph.com/issues/44722 into a backport tracker linked to https://tracker.ceph.com/issues/44733 and added

`Backport: octopus, nautilus`

We can pretend that the nautilus PR is a backport of this one, and do the octopus backport by the book. OK?

**Comment by ivancich** [comment]
> We can pretend that the nautilus PR is a backport of this one, and do the octopus backport by the book. OK?

Sounds good to me, @smithfarm .

**Comment by tchaikov** [comment]
adding needs-qa -- once the the builders are happy with this change, i think it's good to merge.

**Comment by smithfarm** [comment]
@ivancich Can you add `Fixes: https://tracker.ceph.com/issues/44733` to the commit message?

**Comment by ivancich** [comment]
QA results:

http://pulpito.ceph.com/ivancich-2020-05-02_19:31:41-rgw-octopus-rados-list-dev-distro-basic-smithi/
http://pulpito.ceph.com/ivancich-2020-05-02_19:31:18-rgw-octopus-rados-list-dev-distro-basic-smithi/

No QA failures relate to the added code. 

Note 1: new QA test for the tool is here:
http://pulpito.ceph.com/ivancich-2020-05-02_19:31:41-rgw-octopus-rados-list-dev-distro-basic-smithi/5016572/

Note 2: I need to come back with another pass to get rgw-orphan-list testing for versioned buckets.

**Comment by ivancich** [comment]
> if we're backporting this for nautilus/octopus and it gets enough testing in the field, we can consider removing the 'experimental' warnings for pacific

Agreed!

**Comment by ivancich** [comment]
@smithfarm The master PR merged. So this is now ready for further review and QA. If you would glance over it and make sure there are no clerical or process issues given its non-standard history, I'd appreciate it.

See: https://github.com/ceph/ceph/pull/34148

File: src/rgw/rgw_orphan.cc [hunk 1]
```diff
@@ -57,6 +59,30 @@ static string obj_fingerprint(const string& oid, const char *force_ns = NULL)
   return s.substr(0, i + 1);
 }
 
+
+static string obj_force_ns(const string& oid, const char *force_ns)
```
Comment (cbodley) [inline src/rgw/rgw_orphan.cc:63]: getting compiler warnings from this:
```
src/rgw/rgw_orphan.cc:63:15: warning: 'std::string obj_force_ns(const string&, const char*)' defined but not used [-Wunused-function]
   63 | static string obj_force_ns(const string& oid, const char *force_ns)
      |               ^~~~~~~~~~~~
```

**Pull request merged**